Restores goal speed, excluding hoops and dropshot

Timmeey86 opened the pull request:
Proposed fix for issue #2 

How it works:
- We retrieve the Y coordinate for every goal on the map. As soon as the ball transitions beyond that value (in absolute coordinates), we switch a bool flag to remember the ball is in the goal, and display the goal speed. Afterwards, we ignore any further event until the ball is no longer in the goal (the way it gets out of there is not relevant, luckily)
- Unfortunately, this does not work for drop shot or hoops. We could make it work for drop shot by doing the same with the Z coordinate probably, but it would be a lot more complicated for hoops (round goal, hovering over the playing field)

Performed tests:
- Does not trigger anything in hoops
- Does not trigger anything in dropshot
- Does not influence custom workshop maps which don't have goals
- Works with goal scoring enabled
- Works with goal scoring disabled
- Works in free play
- Works in custom training
- Works in snow day
- Works in casual
- Works when changing map while the ball is in the goal

Potential issues:
- The range-based for loop causes warning C4267 (size_t to int conversion). If you want, you can disable the warning or figure out a different loop which does not trigger it
- Currently, the check for a goal/non-goal happens on every tick. I can still get 240fps with this so it seems quick enough. If this bother you, you might want to find a different event to hook to, which triggers not as often, but still often enough

gobbles86 committed:
Restores goal speed, excluding hoops and dropshot

gobbles86 committed:
Removes debugging output

rivques commented:
Just one really nitpicky issue I see with this: Shouldn't instead of watching to see if the ball's y value goes beyond the y value of the goal, you should be watching for the y value of the ball to exceed the y value of the goal *plus the ball's radius*?
Other than that, tysm for fixing this!

Timmeey86 commented:
You are right, I wrongly assumed a goal would occur if half of the ball is behind the line, but looks like the whole ball has to be behind it

CinderBlocc commented:
Might be good to take the ball's position relative to the goal's location and normal instead of just the goal's Y location. Certain workshop maps like the multigoal map wont necessarily have their goal aligned with the Y axis. Also I would recommend leaving the events in there. I'm assuming they only dont fire when the goals are disabled by BM. They theoretically should still work if the goals are enabled, or in non-freeplay modes.

gobbles86 committed:
Restores events and considers ball radius

The events are necessary since we don't always get a tick before a goal explosion when goal scoring is enabled.
The ball radius needs to be considered since a goal only "counts" if the whole ball is beyond the goal position

Timmeey86 commented:
I was about to restore the events anyway (I pushed it now), since it turned out that with very fast shots, you don't always get a tick before the ball explodes, when goal scoring is turned on.
About the multigoal map: Sorry, never heard of that one :-D
I'm afraid my mathematical knowledge is quite limited regarding 3D space. Is it possible to either merge this and create a new issue, or find someone who can implement and test the multi goal map?

gobbles86 committed:
Excludes more game modes

Goals can not as easily be detected in these game modes compared to Soccar maps

Timmeey86 commented:
Looks like I need to exclude rumble and snow day as well, since some maps have their goal within the arena bounds.
I updated the code to exclude workshop maps as well.
In the current state, goal shot speeds can only be detected on regular soccar maps, but I believe this is what most users are looking for anyway. Merging this would give you the option to make it work for the majority, then take your time working out something for the edge cases.

CinderBlocc commented:
Is the original issue only occurring during freeplay when bakkesmod has the goals disabled? If so, I think excluding those other playlist IDs wont resolve the issue directly since those aren't freeplay. I haven't tested this at all so I genuinely don't know when the issue occurs.

Timmeey86 commented:
The original issue is in free play, in any kind of map (soccar, rumble, etc), in case goal scoring is turned off in BM
This PR fixes any Soccar Free Play maps, so what's left broken is free play rumble, free play drop shot etc, all in case goal scoring is turned off. Anything that has a goal explosion, i.e. casual and ranked will be unaffected, since I restored the event hooks

CinderBlocc commented:
Sounds good, I'll merge and ask one of the BM mods to submit the fixed version to bakkesplugins. Thanks for the help!
